src/rust/engine/fs/src/lib.rs

```diff
@@ -675,34 +678,51 @@ impl PosixFS {
     self.ignore.is_ignored(stat)
   }
 
-  pub fn read_file(&self, file: &File) -> impl Future<Item = FileContent, Error = io::Error> {
+  ///
+  /// TODO: See the note on references in ASYNC.md.
+  ///
+  pub fn read_file<'a, 'b, 'c>(
+    &'a self,
+    file: &'b File,
+  ) -> BoxFuture<'c, Result<FileContent, io::Error>> {
     let path = file.path.clone();
     let path_abs = self.root.0.join(&file.path);
-    self.executor.spawn_on_io_pool(future::lazy(move || {
-      let is_executable = path_abs.metadata()?.permissions().mode() & 0o100 == 0o100;
-      std::fs::File::open(&path_abs)
-        .and_then(|mut f| {
-          let mut content = Vec::new();
-          f.read_to_end(&mut content)?;
-          Ok(FileContent {
-            path: path,
-            content: Bytes::from(content),
-            is_executable,
-          })
-        })
-        .map_err(|e| {
-          io::Error::new(
-            e.kind(),
-            format!("Failed to read file {:?}: {}", path_abs, e),
-          )
+    let executor = self.executor.clone();
+    Box::pin(async move {
```

Comment:
I guess if it's how it is, it's how it is, but I'm kind of surprised at the amount of boilerplate to run a closure asynchronously - `Box::pin(async move { spawn_blocking( move || {` seems like more boilerplate than I'd expect :(

Reply:
This is the boilerplate required to "box" an async block (note: not a closure). The `.boxed()` combinator was reintroduced, so I think that this could maybe also be written:
```
let f = async move { .. };
f.boxed()
```
...but chaining a combinator off a block requires a the local variable... etc.

See the note in `ASYNC.md` (and all of the callsites in the patch that reference it: this one included): the explicit use of `BoxFuture` is a stopgap to improve the `futures 0.1`/stdlib futures boundary.